File: android/src/main/java/expo/modules/ecsdkexpo/EcsdkExpoModule.kt
package expo.modules.ecsdkexpo

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL
import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import expo.modules.kotlin.Promise

import com.elerts.ecsdk.ECSDK
import com.elerts.ecsdk.api.model.ECUserData
import com.elerts.ecsdk.api.model.ECClientData
import com.elerts.ecsdk.ui.model.ECUIClientData
import com.elerts.ecsdk.utils.ECOrganizationHelper
import com.elerts.ecsdk.ui.utility.ECPhoneHelper
import com.elerts.ecsdk.api.ECAPIListener
import com.elerts.ecsdk.api.model.ECError
import com.elerts.ecsdk.utils.ECUtils
import com.elerts.ecsdk.utils.ECPreferenceManager
import com.elerts.ecsdk.ui.activity.ECProfileActivity
import com.elerts.ecsdk.api.model.organization.ECOrganizationData
import com.elerts.ecsdk.ui.ECUISDK
import com.elerts.ecsdk.ui.activity.ECReportActivity
import com.elerts.ecsdk.ui.activity.ECAddOrgActivity
import com.elerts.ecsdk.ui.notification.ECNotificationHelper
import com.google.gson.JsonArray
import com.google.firebase.messaging.FirebaseMessaging

class EcsdkExpoModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('EcsdkExpo')` in JavaScript.
    Name("EcsdkExpo")

    // Defines constant property on the module.
    Constant("PI") {
      Math.PI
    }

     // ECSDK Integration Methods
    
    // Initialize the ECSDK with configuration
    Function("initializeSDK") { config: Map<String, Any?> ->
      // Note: Actual ECSDK initialization should be done in the Application class
      // This method can be used to update configuration if needed
      // The actual initialization code should be in ECSDKApplication.kt
    }

    // Create a new client account
    AsyncFunction("createClient") { userData: Map<String, Any?>, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        val regData = ECUserData(context).apply {
          deviceId = userData["deviceId"] as? String ?: ECUtils.getDeviceId(context)
          firstName = userData["firstName"] as? String
          lastName = userData["lastName"] as? String
          email = userData["email"] as? String
          phone = userData["phone"] as? String
        }
        
        ECSDK.register(context, object : ECAPIListener<ECClientData> {
          override fun onAPICompleted(clientData: ECClientData?) {
            if (clientData != null && clientData.token != null) {
              promise.resolve(mapOf("token" to clientData.token))
            } else {
              promise.reject("ECSDK_ERROR", "Failed to register: no client data", null)
            }
          }
          
          override fun onAPIProgress(bytesUploaded: Long, totalBytes: Long) {}
          
          override fun onAPIError(error: ECError) {
            promise.reject("ECSDK_ERROR", error.errorMessage ?: "Unknown error", null)
          }
        }, regData)
        
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Update client information
    AsyncFunction("updateClientInfo") { userData: Map<String, Any?>, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        val regData = ECUserData(context).apply {
          firstName = userData["firstName"] as? String
          lastName = userData["lastName"] as? String
          email = userData["email"] as? String
          phone = userData["phone"] as? String
        }
        
        val clientData = ECUIClientData(context)
        
        ECSDK.clientUpdate(context, object : ECAPIListener<ECUserData> {
          override fun onAPICompleted(result: ECUserData?) {
            promise.resolve(null)
          }
          
          override fun onAPIProgress(bytesUploaded: Long, totalBytes: Long) {}
          
          override fun onAPIError(error: ECError) {
            promise.reject("ECSDK_ERROR", error.errorMessage ?: "Unknown error", null)
          }
        }, regData, clientData)
        
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Get active organization
    AsyncFunction("getActiveOrganization") { promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        val activeOrg = ECOrganizationHelper.getActiveOrg(context)
        if (activeOrg != null) {
          promise.resolve(mapOf(
            "organizationId" to activeOrg.id,
            "name" to activeOrg.name,
            "description" to ""
          ))
        } else {
          promise.resolve(mapOf(
            "organizationId" to 0,
            "name" to "",
            "description" to ""
          ))
        }
        
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Show call police prompt
    Function("showCallPrompt") {
      val activity = appContext.currentActivity ?: throw Exception("No current activity")
      ECPhoneHelper.showCallPrompt(activity)
    }


    // Present organization screen
    Function("presentOrganizationScreen") {
      val activity = appContext.currentActivity ?: throw Exception("No current activity")
      val organizations = Intent(activity, ECAddOrgActivity::class.java)
      activity.startActivity(organizations)
    }

    // Present profile screen
    Function("presentProfileScreen") {
      val activity = appContext.currentActivity ?: throw Exception("No current activity")
      val profile = Intent(activity, ECProfileActivity::class.java)
      activity.startActivity(profile)
    }

    // List organizations
    AsyncFunction("listOrganizations") { joined: Boolean, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        ECOrganizationHelper.apiOrganizationList(context, joined, object : ECAPIListener<Map<String, List<ECOrganizationData>>> {
          override fun onAPICompleted(result: Map<String, List<ECOrganizationData>>?) {
            if (result != null) {
              val organizations = result["orgs"]
              val orgList = organizations?.map { org ->
                mapOf(
                  "organizationId" to org.id,
                  "name" to org.name,
                  "description" to ""
                )
              } ?: emptyList()
              promise.resolve(orgList)
            } else {
              promise.resolve(emptyList<Map<String, Any?>>())
            }
          }
          
          override fun onAPIProgress(bytesUploaded: Long, totalBytes: Long) {}
          
          override fun onAPIError(error: ECError) {
            promise.reject("ECSDK_ERROR", error.errorMessage ?: "Failed to get organizations", null)
          }
        })
        
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Join organization
    AsyncFunction("joinOrganization") { organizationId: Int, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        val clientData = ECUIClientData(context)
        
        // First get all organizations to find the ones matching the IDs
        ECOrganizationHelper.apiOrganizationList(context, true, object : ECAPIListener<Map<String, List<ECOrganizationData>>> {
          override fun onAPICompleted(result: Map<String, List<ECOrganizationData>>?) {
            if (result != null) {
              val allOrgs = result["orgs"] ?: emptyList()
              val orgToJoin = allOrgs.find { org -> org.id.toString() == organizationId.toString() }
              
              if (orgToJoin != null) {
                ECSDK.joinOrganizations(context, object : ECAPIListener<JsonArray> {
                  override fun onAPICompleted(result: JsonArray?) {
                    // Set the org as active
                    ECOrganizationHelper.setActiveOrg(context, orgToJoin)
                    promise.resolve(mapOf("success" to true))
                  }
                  
                  override fun onAPIProgress(bytesUploaded: Long, totalBytes: Long) {}
                  
                  override fun onAPIError(error: ECError) {
                    promise.reject("ECSDK_ERROR", error.errorMessage ?: "Failed to join organization", null)
                  }
                }, clientData, listOf(orgToJoin))
              } else {
                promise.reject("ECSDK_ERROR", "Organization not found", null)
              }
            } else {
              promise.reject("ECSDK_ERROR", "Failed to get organizations", null)
            }
          }
          
          override fun onAPIProgress(bytesUploaded: Long, totalBytes: Long) {}
          
          override fun onAPIError(error: ECError) {
            promise.reject("ECSDK_ERROR", error.errorMessage ?: "Failed to get organizations", null)
          }
        })
        
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Get available organizations
    Function("getAvailableOrganizations") {
      val context = appContext.reactContext ?: throw Exception("React context not available")
      
      val joinedOrgs = ECOrganizationHelper.getAppOrgs(context)
      if (joinedOrgs != null && joinedOrgs.isNotEmpty()) {
        joinedOrgs.map { org ->
          mapOf(
            "organizationId" to org.id,
            "name" to org.name,
            "description" to ""
          )
        }
      } else {
        emptyList<Map<String, Any?>>()
      }
    }

    // Set active organization
    AsyncFunction("setActiveOrganization") { organizationId: Int, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        // First try to find in joined organizations
        val joinedOrgs = ECOrganizationHelper.getAppOrgs(context)
        val targetOrgFromJoined = joinedOrgs?.find { it.id == organizationId }
        
        if (targetOrgFromJoined != null) {
          ECOrganizationHelper.setActiveOrg(context, targetOrgFromJoined)
          promise.resolve(mapOf("success" to true))
        } else {
          // If not found in joined orgs, search in all available orgs
          ECOrganizationHelper.apiOrganizationList(context, true, object : ECAPIListener<Map<String, List<ECOrganizationData>>> {
            override fun onAPICompleted(result: Map<String, List<ECOrganizationData>>?) {
              if (result != null) {
                val organizations = result["orgs"]
                val targetOrg = organizations?.find { it.id == organizationId }
                
                if (targetOrg != null) {
                  ECOrganizationHelper.setActiveOrg(context, targetOrg)
                  promise.resolve(mapOf("success" to true))
                } else {
                  promise.reject("ECSDK_ERROR", "Organization not found", null)
                }
              } else {
                promise.reject("ECSDK_ERROR", "Failed to get organizations", null)
              }
            }
            
            override fun onAPIProgress(bytesUploaded: Long, totalBytes: Long) {}
            
            override fun onAPIError(error: ECError) {
              promise.reject("ECSDK_ERROR", error.errorMessage ?: "Failed to get organizations", null)
            }
          })
        }
        
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Present report screen
    Function("presentReportScreen") {
      val activity = appContext.currentActivity ?: throw Exception("No current activity")
      val context = appContext.reactContext ?: throw Exception("React context not available")
      
      // Check if there's an active organization
      val activeOrg = ECOrganizationHelper.getActiveOrg(context)
      
      // Warn user if no active org is selected (don't auto-select)
      if (activeOrg == null) {
        val joinedOrgs = ECOrganizationHelper.getAppOrgs(context)
        if (joinedOrgs != null && joinedOrgs.isNotEmpty()) {
          throw Exception("No active organization selected. Please select an organization from your joined organizations list before submitting a report.")
        } else {
          throw Exception("No active organization. Please join an organization first.")
        }
      }
      
      // Create Intent as per SDK documentation
      val reportIntent = Intent()
      reportIntent.setClass(activity, ECReportActivity::class.java)
      activity.startActivity(reportIntent)
    }

    // Show message list activity
    Function("showMessageList") {
      val activity = appContext.currentActivity ?: throw Exception("No current activity")
      val messageListIntent = Intent(activity, ECUISDK.getMessageListClass())
      activity.startActivity(messageListIntent)
    }

    // FCM Token Management
    
    // Get the current FCM token
    AsyncFunction("getFCMToken") { promise: Promise ->
      try {
        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
          if (task.isSuccessful) {
            val token = task.result
            if (token != null) {
              promise.resolve(token)
            } else {
              promise.reject("ECSDK_ERROR", "FCM token is null", null)
            }
          } else {
            promise.reject("ECSDK_ERROR", "Failed to get FCM token: ${task.exception?.message}", task.exception)
          }
        }
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Update FCM token with ELERTS server
    AsyncFunction("updateFCMToken") { fcmToken: String?, promise: Promise ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        // Save token to preferences
        if (fcmToken != null) {
          ECPreferenceManager.putString(context, ECSDKFirebaseMessagingService.PROPERTY_REG_ID, fcmToken)
        }
        
        // Create user data with device ID
        val regData = ECUserData(context).apply {
          deviceId = ECUtils.getDeviceId(context)
        }
        
        // Get client data
        val clientData = ECClientData(context)
        
        // Only update if we have a client token (user is registered)
        if (clientData.token != null) {
          ECSDK.clientUpdate(context, object : ECAPIListener<ECUserData> {
            override fun onAPICompleted(result: ECUserData?) {
              promise.resolve(mapOf("success" to true))
            }
            
            override fun onAPIProgress(bytesUploaded: Long, totalBytes: Long) {}
            
            override fun onAPIError(error: ECError) {
              promise.reject("ECSDK_ERROR", error.errorMessage ?: "Failed to update FCM token", null)
            }
          }, regData, clientData)
        } else {
          promise.reject("ECSDK_ERROR", "Client not registered. Please register first.", null)
        }
        
      } catch (e: Exception) {
        promise.reject("ECSDK_ERROR", e.message, e)
      }
    }

    // Get stored FCM token from preferences
    Function("getStoredFCMToken") {
      val context = appContext.reactContext ?: throw Exception("React context not available")
      ECPreferenceManager.getString(context, ECSDKFirebaseMessagingService.PROPERTY_REG_ID, null)
    }

    // Register for remote notifications (FCM)
    // Note: FCM auto-registers when Firebase initializes, but this method
    // can be used to manually trigger token refresh for consistency with iOS API
    Function("registerForRemoteNotifications") {
      // On Android, FCM automatically handles registration via ECSDKFirebaseMessagingService
      // This method triggers a token refresh which will call onNewToken() in the service
      // The service will automatically send the token to ELERTS server if user is registered
      FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
        if (task.isSuccessful) {
          val token = task.result
          android.util.Log.d("EcsdkAndroidExpoModule", "FCM token refreshed: ${token?.take(20)}...")
          // Token will be automatically handled by ECSDKFirebaseMessagingService.onNewToken()
          // if it's a new token, or can be manually updated via updateFCMToken()
        } else {
          android.util.Log.e("EcsdkAndroidExpoModule", "Failed to refresh FCM token: ${task.exception?.message}")
        }
      }
    }

    // User Login/Logout Management
    
    // Login with existing ELERTS client token
    Function("login") { clientToken: String ->
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        // Use ECUISDK.login() as per the SDK documentation
        ECUISDK.login(clientToken)
        
        android.util.Log.d("EcsdkAndroidExpoModule", "User logged in with token")
      } catch (e: Exception) {
        throw Exception("Failed to login: ${e.message}")
      }
    }
    
    // Logout current user
    Function("logout") {
      try {
        val context = appContext.reactContext ?: throw Exception("React context not available")
        
        // Use ECUISDK.logout() as per the SDK documentation
        ECUISDK.logout()
        
        android.util.Log.d("EcsdkAndroidExpoModule", "User logged out")
      } catch (e: Exception) {
        throw Exception("Failed to logout: ${e.message}")
      }
    }
    
    // Get current client token
    Function("getClientToken") {
      val context = appContext.reactContext ?: throw Exception("React context not available")
      val clientData = ECClientData(context)
      clientData.token
    }
  }
}

File: ios/EcsdkExpoModule.swift
import ExpoModulesCore
import ELERTSKitCore
import ELERTSKitUI

// Helper function to extract organization ID from EKOrganization
private func getOrganizationId(from org: EKOrganization) -> Int {
  var orgId: Int = 0
  let mirror = Mirror(reflecting: org)
  for child in mirror.children {
    if let label = child.label {
      // Try common property name variations
      if label == "organizationId" || label == "id" || label == "_organizationId" || 
         label == "organizationID" || label == "_id" || label.contains("organizationId") {
        if let value = child.value as? Int {
          orgId = value
          break
        }
      }
    }
  }
  return orgId
}

// Helper function to convert EKOrganization to dictionary
private func organizationToDict(_ org: EKOrganization) -> [String: Any] {
  return [
    "organizationId": getOrganizationId(from: org),
    "name": org.name,
    "description": org.description
  ]
}

// Helper function to create profile field items
private func createProfileFieldItems() -> [EKUIProfileFieldItem] {
  let firstNameField = EKUIProfileFieldItem(
    name: "First Name",
    type: EKUIProfileItemType.firstName
  )
  
  let lastNameField = EKUIProfileFieldItem(
    name: "Last Name",
    type: EKUIProfileItemType.lastName
  )
  
  let emailField = EKUIProfileFieldItem(
    name: "Email",
    type: EKUIProfileItemType.email
  )
  
  let phoneField = EKUIProfileFieldItem(
    name: "Mobile Number",
    type: EKUIProfileItemType.phone
  )
  
  return [firstNameField, lastNameField, emailField, phoneField]
}

// Helper function to create profile view controller config
private func createProfileVCConfig() -> EKUIProfileConfig {
  return EKUIProfileConfig(
    name: "My Info",
    header: nil,
    footer: "Update your profile information",
    isRequired: true,
    allowEmailRegistration: true,
    items: createProfileFieldItems()
  )
}

// Custom profile view controller with custom dismiss buttons
private class CustomProfileViewController: EKUIProfileViewController {
  override func viewDidLoad() {
    super.viewDidLoad()
    
    // Set up custom navigation bar buttons
    setupCustomButtons()
  }
  
  private func setupCustomButtons() {
    // Create cancel button
    let cancelButton = UIBarButtonItem(
      barButtonSystemItem: .cancel,
      target: self,
      action: #selector(cancelButtonTapped)
    )
    self.navigationItem.leftBarButtonItem = cancelButton
    
    // Create save button
    let saveButton = UIBarButtonItem(
      barButtonSystemItem: .save,
      target: self,
      action: #selector(saveButtonTapped)
    )
    self.navigationItem.rightBarButtonItem = saveButton
  }
  
  @objc private func cancelButtonTapped() {
    // Dismiss without saving
    self.dismiss(animated: true, completion: nil)
  }
  
  @objc private func saveButtonTapped() {
    // Save values with completion handler
    self.saveValues { [weak self] (saved) in
      guard let self = self else { return }
      
      // The saveValues method will check for errors and show them
      // It will also save to keychain and update ELERTS Cloud
      // Dismiss the view controller in the callback
      if saved {
        DispatchQueue.main.async {
          self.dismiss(animated: true, completion: nil)
        }
      }
    }
  }
}

public class EcsdkExpoModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('EcsdkExpo')` in JavaScript.
    Name("EcsdkExpo")

    // Defines constant property on the module.
    Constant("PI") {
      Double.pi
    }
    
    // Get client token
    Function("getClientToken") {
      return ELERTSKit.getClientToken()
    }

    // Create a new client account
    AsyncFunction("createClient") { (profileDict: [String: Any], promise: Promise) -> Void in
      let firstName = profileDict["firstName"] as? String ?? ""
      let lastName = profileDict["lastName"] as? String ?? ""
      let email = profileDict["email"] as? String ?? ""
      let phone = profileDict["phone"] as? String ?? ""
      // Note: otherFields is not currently supported as it requires [EKUserProfileField]?
      // Passing nil for now - can be extended later if needed
      
      let profile = EKUserProfile(
        firstName: firstName,
        lastName: lastName,
        email: email,
        phone: phone,
        otherFields: nil
      )
      
      ELERTSKit.createClient(profile: profile) { result in
        switch result {
        case .success(let clientToken):
          promise.resolve(clientToken)
        case .failure(let error):
          promise.reject("CREATE_CLIENT_ERROR", "Error creating account: \(error.localizedDescription)")
        }
      }
    }

    // Join organization
    AsyncFunction("joinOrganization") { (organizationId: Int, promise: Promise) -> Void in
      guard EKKeychain.getString(.token) != nil else {
        promise.reject("JOIN_ORG_ERROR", "No client token available. Please login first.")
        return
      }
      
      let organization = EKOrganization(organizationId: organizationId)
      ELERTSKit.joinOrganization([organization]) { joinResult in
        switch joinResult {
        case .success:
          promise.resolve(["success": true])
        case .failure(let error):
          promise.reject("JOIN_ORG_ERROR", "Error joining org: \(error.localizedDescription)")
        }
      }
    }

    // List organizations (joined or available)
    AsyncFunction("listOrganizations") { (joined: Bool, promise: Promise) -> Void in
      guard EKKeychain.getString(.token) != nil else {
        promise.reject("LIST_ORG_ERROR", "No client token available. Please login first.")
        return
      }
      
      // Call the SDK's API to list organizations
      // The SDK may automatically update ELERTSKit.organizations internally
      _ = EKAPI.Organization.list(joined: joined).go(EKAPI.Organization.ListResult.self) { result in
        switch result {
        case .success(let listResult):
          // Unwrap optional Set and convert to array of dictionaries
          guard let organizations = listResult.organizations else {
            promise.resolve([])
            return
          }
          
          // Ensure ELERTSKit.organizations is updated with the fetched organizations
          ELERTSKit.organizations.fire(organizations)
          
          let orgs = Array(organizations).map { (org: EKOrganization) -> [String: Any] in
            return organizationToDict(org)
          }
          promise.resolve(orgs)
        case .failure(let error):
          promise.reject("LIST_ORG_ERROR", "Error listing organizations: \(error.localizedDescription)")
        @unknown default:
          promise.reject("LIST_ORG_ERROR", "Unknown error listing organizations")
        }
      }
    }

    // Set active organization
    AsyncFunction("setActiveOrganization") { (organizationId: Int, promise: Promise) -> Void in
      // Try to find the organization in the SDK's organizations list first
      // This ensures we use the full organization object with all its data
      var organizationToSet: EKOrganization? = nil
      
      if let organizations = ELERTSKit.organizations.lastDataFired {
        organizationToSet = organizations.first { org in
          getOrganizationId(from: org) == organizationId
        }
      }
      
      if organizationToSet == nil {
        organizationToSet = EKOrganization(organizationId: organizationId)
      }
      
      ELERTSKit.activeOrganization.fire(organizationToSet!)
      promise.resolve(["success": true])
    }

    // Get active organization
    AsyncFunction("getActiveOrganization") { (promise: Promise) -> Void in
      if let activeOrg = ELERTSKit.activeOrganization.lastDataFired as? EKOrganization {
        promise.resolve(organizationToDict(activeOrg))
      } else {
        // Return dictionary with organizationId: 0 to indicate no active organization
        promise.resolve([
          "organizationId": 0,
          "name": "",
          "description": ""
        ])
      }
    }

    // Get list of available organizations
    Function("getAvailableOrganizations") {
      if let organizations = ELERTSKit.organizations.lastDataFired {
        // Convert Set to Array if needed
        let orgArray = Array(organizations)
        return orgArray.map { (org: EKOrganization) -> [String: Any] in
          return organizationToDict(org)
        }
      }
      return []
    }

    // Update client info
    AsyncFunction("updateClientInfo") { (profileDict: [String: Any], promise: Promise) -> Void in
      let firstName = profileDict["firstName"] as? String ?? ""
      let lastName = profileDict["lastName"] as? String ?? ""
      let email = profileDict["email"] as? String ?? ""
      let phone = profileDict["phone"] as? String ?? ""
      // Note: otherFields is not currently supported as it requires [EKUserProfileField]?
      // Passing nil for now - can be extended later if needed
      
      let profile = EKUserProfile(
        firstName: firstName,
        lastName: lastName,
        email: email,
        phone: phone,
        otherFields: nil
      )
      
      ELERTSKit.updateClientInfo(profile: profile) { result in
        switch result {
        case .success:
          promise.resolve(true)
        case .failure(let error):
          promise.reject("UPDATE_CLIENT_ERROR", "Error updating account: \(error.localizedDescription)")
        }
      }
    }

    // Login with client token
    Function("login") { (clientToken: String) -> Void in
      ELERTSKitUI.login(elertsClientToken: clientToken)
    }

    // Logout
    Function("logout") {
      ELERTSKitUI.logout()
    }

    // Present report screen
    Function("presentReportScreen") {
      DispatchQueue.main.async {
        // Get the root view controller
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
          return
        }
        
        // Get the topmost view controller
        var topViewController = rootViewController
        while let presented = topViewController.presentedViewController {
          topViewController = presented
        }
        
        // Check if active organization is set
        guard let activeOrg = ELERTSKit.activeOrganization.lastDataFired as? EKOrganization else {
          let alert = UIAlertController(
            title: "No Active Organization",
            message: "Please select and join an organization before reporting a problem.",
            preferredStyle: .alert
          )
          alert.addAction(UIAlertAction(title: "OK", style: .default))
          topViewController.present(alert, animated: true)
          return
        }
        
        // Try to find the full organization object from the SDK's list
        var fullOrg: EKOrganization? = nil
        if let organizations = ELERTSKit.organizations.lastDataFired {
          let orgId = getOrganizationId(from: activeOrg)
          fullOrg = organizations.first { org in
            getOrganizationId(from: org) == orgId
          }
        }
        
        // Use the full organization object if found, otherwise use the active one
        let orgToUse = fullOrg ?? activeOrg
        ELERTSKit.activeOrganization.fire(orgToUse)
        
        // Create and present the report view controller
        let reportVC = EKUIReportViewController.initInNavigationController(reportConfig: EKUIReportConfig.defaultConfig())
        topViewController.present(reportVC, animated: true, completion: nil)
      }
    }

    // Present organization screen
    Function("presentOrganizationScreen") {
      DispatchQueue.main.async {
        // Get the root view controller
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
          return
        }
        
        // Get the topmost view controller
        var topViewController = rootViewController
        while let presented = topViewController.presentedViewController {
          topViewController = presented
        }
        
        // Check if client token is available
        guard EKKeychain.getString(.token) != nil else {
          let alert = UIAlertController(
            title: "Not Logged In",
            message: "Please login first before managing organizations.",
            preferredStyle: .alert
          )
          alert.addAction(UIAlertAction(title: "OK", style: .default))
          topViewController.present(alert, animated: true)
          return
        }
        
        // Create and present the organization view controller
        let organizationVC = EKUIOrganizationViewController(
          config: EKUIOrgFeedConfig(
            name: "",
            header: nil,
            footer: nil,
            orgFeedItem: EKUIOrgFeedItem(name: "", fieldName: "")
          )
        )
        let navController = UINavigationController(rootViewController: organizationVC)
        topViewController.present(navController, animated: true, completion: nil)
      }
    }

    // Present profile screen
    Function("presentProfileScreen") {
      DispatchQueue.main.async {
        // Get the root view controller
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
          return
        }
        
        // Get the topmost view controller
        var topViewController = rootViewController
        while let presented = topViewController.presentedViewController {
          topViewController = presented
        }
        
        // Check if client token is available
        guard EKKeychain.getString(.token) != nil else {
          let alert = UIAlertController(
            title: "Not Logged In",
            message: "Please login first before updating your profile.",
            preferredStyle: .alert
          )
          alert.addAction(UIAlertAction(title: "OK", style: .default))
          topViewController.present(alert, animated: true)
          return
        }
        
        // Create profile config
        let userProfileConfig = createProfileVCConfig()
        
        // Create custom profile view controller with config
        let userProfileVC = CustomProfileViewController(config: userProfileConfig)
        
        // Wrap in navigation controller
        let navController = UINavigationController(rootViewController: userProfileVC)
        
        // Present the profile view controller
        topViewController.present(navController, animated: true, completion: nil)
      }
    }

    // Show call police prompt
    Function("showCallPrompt") {
      DispatchQueue.main.async {
        // Get active organization to retrieve phone info (optional, for reference)
        // The phone and phoneTitle can be used for logging or custom implementation if needed
        if let activeOrg = ELERTSKit.activeOrganization.lastDataFired as? EKOrganization {
          let _ = activeOrg.phone
          let _ = activeOrg.phoneTitle
        }
        
        // Show the call alert using EKCallHelper
        // customNumber: nil (uses default police number)
        // panicSent: nil (no panic callback - SDK will log to console when user calls)
        // showOrgName: false (don't show organization name in alert)
        EKCallHelper.showCallAlert(customNumber: nil, panicSent: nil, showOrgName: false)
      }
    }

    // Show message list
    Function("showMessageList") {
      DispatchQueue.main.async {
        // Get the root view controller
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
          return
        }
        
        // Get the topmost view controller
        var topViewController = rootViewController
        while let presented = topViewController.presentedViewController {
          topViewController = presented
        }
        
        // Check if client token is available
        guard EKKeychain.getString(.token) != nil else {
          let alert = UIAlertController(
            title: "Not Logged In",
            message: "Please login first before viewing messages.",
            preferredStyle: .alert
          )
          alert.addAction(UIAlertAction(title: "OK", style: .default))
          topViewController.present(alert, animated: true)
          return
        }
        
        // Create and present the message list view controller
        // EKUIThreadListViewController displays the list of messages
        // When a user taps on a message, the SDK will automatically handle navigation
        // to EKUIThreadMessageViewController for message detail
        let messageListVC = EKUIThreadListViewController()
        let navController = UINavigationController(rootViewController: messageListVC)
        topViewController.present(navController, animated: true, completion: nil)
      }
    }

    // Register for remote notifications
    Function("registerForRemoteNotifications") {
      EKNotificationManager.registerForRemoteNotification()
    }
  }
}
